### PR TITLE
Add -f/-u arguments to ParseDb merge

### LINKED ISSUE #167: Add sample annotation argument to ParseDb-merge
**[Original report](https://bitbucket.org/kleinstein/changeo/issues/167) by Jason Vander Heiden (Bitbucket: [javh](https://bitbucket.org/javh), GitHub: [javh](https://github.com/javh)).**

----------------------------------------

ParseDb-merge could use arguments to annotate the merged files with sample identifiers. Eg, `-f <field>` and `-u <samples>`.

### PR DESCRIPTION
**Pull request by ssnn-airr**
Adds -f/-u arguments to the merge subcommand, allowing merged records to be annotated with a per-input-file sample identifier field.

Fixes #167

### CODE REVIEW

File: NEWS.rst [hunk 1]
```diff
@@ -18,10 +27,10 @@ MakeDb:
 
 Gene:
 
-+ Added support for dual-locus ``TRA/TRD`` gene detection. ``getLocus`` now 
-  correctly identifies genes with the ``TRAV.../DV...`` naming convention 
++ Added support for dual-locus ``TRA/TRD`` gene detection. ``getLocus`` now
```
Comment (cfsullivan16) [inline NEWS.rst:30]: Note: lines 30-33 just whitespace changes; do not actually have to do with this update

File: bin/ParseDb.py [hunk 2]
```diff
@@ -814,6 +819,9 @@ def mergeDbFiles(db_files, drop=False, out_file=None, out_args=default_out_args)
     log['COMMAND'] = 'merge'
     log['FILES'] = ','.join([os.path.basename(f) for f in db_files])
     log['DROP'] = drop
+    if field is not None:
```
Comment (cfsullivan16) [inline bin/ParseDb.py:826]: Should there be a warning if a user enters values but not a field that the entered values will be ignored? Should there be a default field value in this case?

File: bin/ParseDb.py [hunk 3]
```diff
@@ -829,6 +837,8 @@ def mergeDbFiles(db_files, drop=False, out_file=None, out_args=default_out_args)
         field_set = set.union(*map(set, field_list))
     field_order = OrderedDict([(f, None) for f in chain(*field_list)])
     out_fields = [f for f in field_order if f in field_set]
+    if field is not None and field not in out_fields:
```
Comment (cfsullivan16) [inline bin/ParseDb.py]: Should there be a warning about over-writing an existing column if field IS in the out_fields?

File: tests/test_ParseDb.py [hunk 4]
```diff
@@ -25,11 +28,45 @@ class Test_ParseDb(unittest.TestCase):
...
+        self.assertEqual(len(records), 3)
+        self.assertNotIn('SAMPLE', records[0])
+
+        # Merge with sample annotation
+        ParseDb.mergeDbFiles([file1, file2], field='SAMPLE', values=['sample1', 'sample2'],
+                              out_file=out_file)
+        records = self._read_tsv(out_file)
+        results = {r['SEQUENCE_ID']: r['SAMPLE'] for r in records}
+        self.assertDictEqual(results, {'A1': 'sample1', 'A2': 'sample1', 'B1': 'sample2'})
+
```
Comment (cfsullivan16) [inline tests/test_ParseDb.py:71]: These are good tests to ensure that function is working with and without the new functionality. Should there be a couple more tests of edge cases to make sure that warnings and errors are raised when they should be? Examples:
- field is entered but no values entered
- values are entered but no field name is entered
- the field is going to over-write existing data

**Review by cfsullivan16 — COMMENTED** [review]
- I think the changes look good and properly address the issues.
- I made a couple comments suggesting additional warning messages that could be warranted ex. new field will overwrite existing data, values entered with no field (can a default be used in this case, or should values be ignored?)
- More tests could be added to ensure parser errors get triggered as expected, but this might just be overkill?
- I opened the branch in a GitHub codespace and ensured 1. the package installs correctly 2. the tests for merge pass (tests for with and without added fields/values).

I can add these little tweaks, or if we think it's unnecessary I think it's OK to merge.
